#!/usr/bin/Rscript

## load packages
library(argparse, quietly = T, warn.conflicts = F)
library(Sierra)

# parameters --------------------------------------------------------------
parser <- ArgumentParser(description='Normalize APA matrix')

parser$add_argument('--bam', 
                    dest='bam',
                    action='store',
                    help='BAM file')
parser$add_argument('--junctions', 
                    dest='junctions',
                    action='store',
                    help='junction file generated by regtools')
parser$add_argument('--reference', 
                    dest='reference',
                    action='store',
                    default='',
                    help='path to the gtf reference')
parser$add_argument('--whitelist', 
                    dest='whitelist',
                    action='store',
                    default='',
                    help='cell barcode whitelist')
parser$add_argument('--nthreads', 
                    dest='nthreads',
                    action='store',
                    type='integer',
                    default=8,
                    help='the number of threads (default: 8)')

args <- parser$parse_args()

## input load once
bamfile <- args[["bam"]] 
junctions.file <- args[["junctions"]] 
reference.file <- args[["reference"]] #reference.file <- '~/scratch/apa/reference/refdata-gex-mm10-2020-A/genes/genes.gtf'
whitelist.bc.file <- args[["whitelist"]]
num_core <- args[["nthreads"]]

peak.output.file <- "Sierra_peak.txt" #paste0(output_dir,'/Sierra_peak.txt')
count.dirs <- "Sierra_counts" #paste0(output_dir, 'Sierra_counts')
#annot_file <- paste0(output_dir, 'Sierra_peak_annotations.txt')

# peak calling ------------------------------------------------------------
FindPeaks(output.file = peak.output.file,   # output filename
          gtf.file = reference.file,           # gene model as a GTF file
          bamfile = bamfile,                # BAM alignment filename.
          junctions.file = junctions.file,     # BED filename of splice junctions exising in BAM file. 
          ncores = num_core)                          # number of cores to use

# peak count --------------------------------------------------------------
CountPeaks(peak.sites.file = peak.output.file, 
           gtf.file = reference.file,
           bamfile = bamfile, 
           whitelist.file = whitelist.bc.file,
           output.dir = count.dirs, 
           countUMI = TRUE, 
           ncores = num_core)

# peak annotate -----------------------------------------------------------

# New definitions
# genome <- BSgenome.Mmusculus.UCSC.mm10::BSgenome.Mmusculus.UCSC.mm10

# AnnotatePeaksFromGTF(peak.sites.file = peak.merge.output.file, 
#                      gtf.file = reference.file,
#                      output.file = annot_file, 
#                      genome = genome)

################## Stop ##############################################
